pallets/messages/src/lib.rs

```diff
@@ -319,90 +319,22 @@ impl<T: Config> Pallet<T> {
+	/// * `Vec<MessageResponse<T::BlockNumber>>`
 	///
-	/// Result is a paginator response of type [`BlockPaginationResponse`].
-	pub fn get_messages_by_schema(
+	/// Result is a vector response of type [`MessageResponse`].
+	pub fn get_messages_by_schema_and_block(
 		schema_id: SchemaId,
-		pagination: BlockPaginationRequest<T::BlockNumber>,
-	) -> Result<
-		BlockPaginationResponse<T::BlockNumber, MessageResponse<T::BlockNumber>>,
-		DispatchError,
-	> {
-		ensure!(pagination.validate(), Error::<T>::InvalidPaginationRequest);
-
-		let schema = T::SchemaProvider::get_schema_by_id(schema_id);
-		ensure!(schema.is_some(), Error::<T>::InvalidSchemaId);
-
-		let payload_location = schema.unwrap().payload_location;
-		let mut response = BlockPaginationResponse::new();
-		let from: u32 = pagination
-			.from_block
-			.try_into()
-			.map_err(|_| Error::<T>::TypeConversionOverflow)?;
-		let to: u32 =
-			pagination.to_block.try_into().map_err(|_| Error::<T>::TypeConversionOverflow)?;
-		let mut from_index = pagination.from_index;
-
-		'loops: for bid in from..to {
-			let block_number: T::BlockNumber = bid.into();
-			let list = <Messages<T>>::get(block_number, schema_id).into_inner();
-
-			let list_size: u32 =
-				list.len().try_into().map_err(|_| Error::<T>::TypeConversionOverflow)?;
-			for i in from_index..list_size {
-				let m = list[i as usize].clone();
-				response.content.push(m.map_to_response(block_number, &payload_location));
-
-				if Self::check_end_condition_and_set_next_pagination(
-					block_number,
-					i,
-					list_size,
-					&pagination,
-					&mut response,
-				) {
-					break 'loops
-				}
-			}
-
-			// next block starts from 0
-			from_index = 0;
-		}
-
-		Ok(response)
-	}
-
-	/// Checks the end condition for paginated query and set the `PaginationResponse`
-	///
-	/// Returns `true` if page is filled
-	fn check_end_condition_and_set_next_pagination(
+		schema_payload_location: PayloadLocation,
 		block_number: T::BlockNumber,
-		current_index: u32,
-		list_size: u32,
-		request: &BlockPaginationRequest<T::BlockNumber>,
-		result: &mut BlockPaginationResponse<T::BlockNumber, MessageResponse<T::BlockNumber>>,
-	) -> bool {
-		if result.content.len() as u32 == request.page_size {
-			let mut next_block = block_number;
-			let mut next_index = current_index + 1;
-
-			// checking if it's end of current list
-			if next_index == list_size {
-				next_block = block_number + T::BlockNumber::one();
-				next_index = 0;
-			}
-
-			if next_block < request.to_block {
-				result.has_next = true;
-				result.next_block = Some(next_block);
-				result.next_index = Some(next_index);
-			}
-			return true
-		}
-
-		false
+	) -> Vec<MessageResponse<T::BlockNumber>> {
+		<Messages<T>>::get(block_number, schema_id)
+			.into_inner()
+			.iter()
+			.map(|msg| msg.map_to_response(block_number, schema_payload_location))
+			.collect()
```

Comment:
New logic just extracts the `get` from state to expose it through the runtime api